src/librustc_mir/hair/pattern/const_to_pat.rs

```diff
@@ -0,0 +1,406 @@
+        fulfillment_cx.register_bound(
+            &self.infcx, ty::ParamEnv::empty(), adt_ty, structural_teq_def_id, cause);
+
+        // We deliberately skip *reporting* fulfillment errors (via
+        // `report_fulfillment_errors`), for two reasons:
+        //
+        // 1. The error messages would mention `std::marker::StructuralPartialEq`
+        //    (a trait which is solely meant as an implementation detail
+        //    for now), and
+        //
+        // 2. We are sometimes doing future-incompatibility lints for
+        //    now, so we do not want unconditional errors here.
+        fulfillment_cx.select_all_or_error(&self.infcx).is_ok()
+    }
+
+    /// This method traverses the structure of `ty`, trying to find an
+    /// instance of an ADT (i.e. struct or enum) that does not implement
+    /// the `Structural` trait.
+    ///
+    /// The "structure of a type" includes all components that would be
+    /// considered when doing a pattern match on a constant of that
+    /// type.
+    ///
+    ///  * This means this method descends into fields of structs/enums,
+    ///    and also descends into the inner type `T` of `&T` and `&mut T`
+    ///
+    ///  * The traversal doesn't dereference unsafe pointers (`*const T`,
+    ///    `*mut T`), and it does not visit the type arguments of an
+    ///    instantiated generic like `PhantomData<T>`.
+    ///
+    /// The reason we do this search is Rust currently require all ADT's
+    /// reachable from a constant's type to implement `Structural`, a
+    /// trait which essentially says that the implementation of
+    /// `PartialEq::eq` behaves *equivalently* to a comparison against
+    /// the unfolded structure.
+    ///
+    /// For more background on why Rust has this requirement, and issues
+    /// that arose when the requirement was not enforced completely, see
+    /// Rust RFC 1445, rust-lang/rust#61188, and rust-lang/rust#62307.
+    fn search_for_adt_without_structural_match(&self, ty: Ty<'tcx>) -> Option<&'tcx AdtDef> {
+        let mut search = Search {
+            const_to_pat: self,
+            found: None,
+            seen: FxHashSet::default()
+        };
+
+        ty.visit_with(&mut search);
+        return search.found;
+    }
+}
+
+struct Search<'a, 'tcx> {
+    const_to_pat: &'a ConstToPat<'a, 'tcx>,
+
+    // records the first ADT we find that does not implement `Structural`.
+    found: Option<&'tcx AdtDef>,
+
+    // tracks ADT's previously encountered during search, so that
+    // we will not recur on them again.
+    seen: FxHashSet<&'tcx AdtDef>,
+}
+
+impl<'a, 'tcx> TypeVisitor<'tcx> for Search<'a, 'tcx> {
+    fn visit_ty(&mut self, ty: Ty<'tcx>) -> bool {
+        debug!("Search visiting ty: {:?}", ty);
+
+        let (adt_def, substs) = match ty.kind {
+            ty::Adt(adt_def, substs) => (adt_def, substs),
+            ty::RawPtr(..) => {
+                // structural-match ignores substructure of
```

Comment:
This case could use some elaboration.

Reply:
I expanded the comment with an example, hopefully that suffices.

Reply:
Yep, it is great, thanks!